src/number.cr

```diff
@@ -168,6 +168,241 @@ struct Number
     self > other ? 1 : (self < other ? -1 : 0)
   end
 
+  # Prints this number as a `String` using a customizable format.
+  #
+  # *separator* is used as decimal separator, *delimiter* as thousands
+  # delimiter between batches of *group* digits.
+  #
+  # If *decimal_places* is `nil`, all significant decimal places are printed
+  # (similar to `#to_s`). If the argument has a numeric value, the number of
+  # visible decimal places will be fixed to that amount.
+  #
+  # Trailing zeros are omitted if *only_significant* is `true`.
+  #
+  # ```
+  # 123_456.789.format                                            # => "123,456.789"
+  # 123_456.789.format(',', '.')                                  # => "123.456,789"
+  # 123_456.789.format(decimal_places: 2)                         # => "123,456.79"
+  # 123_456.789.format(decimal_places: 6)                         # => "123,456.789000"
+  # 123_456.789.format(decimal_places: 6, only_significant: true) # => "123,456.789"
+  # ```
+  def format(io : IO, separator = '.', delimiter = ',', decimal_places : Int? = nil, *, group : Int = 3, only_significant : Bool = false) : Nil
+    number = self
+    # TODO: Optimize implementation for Int
+    if decimal_places
+      number = number.round(decimal_places)
+    end
+    string = number.abs.to_s
+    integer, _, decimals = string.partition('.')
+
+    int_size = integer.size
+    dec_size = decimals.size
+
+    io << '-' if self < 0
+
+    start = int_size % group
+    start += group if start == 0
+    io.write integer.to_slice[0, start]
+
+    while start < int_size
+      io << delimiter
+      io.write integer.to_slice[start, group]
+      start += group
+    end
+
+    decimal_places ||= dec_size
+
+    if decimal_places > 0
+      io << separator << decimals
+      unless only_significant
+        (decimal_places - dec_size).times do
+          io << '0'
```

Comment:
Just `io << 0` would probably be better.

Reply:
Why do you think it would be better? 😁

Reply:
I was pretty sure just a simple Int would be faster than a Char but nvm:
```Crystal
require "benchmark"

io = IO::Memory.new

Benchmark.ips do |bm|
  bm.report "char" {
    io << '0'
  }

  bm.report "int" {
    io << 0
  }
end
```
```
char  81.07M ( 12.34ns) (± 6.90%)  3 B/op        fastest
 int  76.95M (  13.0ns) (±11.22%)  4 B/op   1.05× slower
```